File: standalone/dzen2-hlwm/ruby/pipehandler.simple.rb
require_relative 'output'

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# pipe

def handle_command_event(monitor, event) 
  # find out event origin
  column = event.split("\t")
  origin = column[0]
    
  tag_cmds = ['tag_changed', 'tag_flags', 'tag_added', 'tag_removed']
  title_cmds = ['window_title_changed', 'focus_changed']

  case origin
  when 'reload'
    os.system('pkill dzen2')
  when 'quit_panel'
    exit
  when *tag_cmds       # splat operator
    set_tag_value(monitor)
  when *title_cmds     # splat operator
    set_windowtitle(column[2])
  end
end

def init_content(monitor, stdin)
  # initialize statusbar before loop
  set_tag_value(monitor)
  set_windowtitle('')
      
  text = get_statusbar_text(monitor)
  stdin.puts(text)
end

def walk_content(monitor, stdin)    
  # start a pipe
  command_in = 'herbstclient --idle'
  
  IO.popen(command_in, "r") do |f|     
    while f do 
      # read next event
      event = f.gets
      handle_command_event(monitor, event)
        
      text = get_statusbar_text(monitor)
      stdin.write(text)
    end
    f.close()    
  end
end

def run_dzen2(monitor, parameters)
  command_out  = 'dzen2 ' + parameters
  IO.popen(command_out, "w") do |f| 
    init_content(monitor, f)
    walk_content(monitor, f) # loop for each event
        
    f.close()    
  end
end

def detach_dzen2(monitor, parameters)
  # warning: Signal.trap is application wide
  Signal.trap("PIPE", "EXIT")
    
  pid = fork { run_dzen2(monitor, parameters) }
  Process.detach(pid)
end

def detach_transset()
  pid = fork do
    sleep(1)
    system('transset .8 -n dzentop >/dev/null')        
  end
    
  Process.detach(pid)
end

File: standalone/lemon-hlwm/php/pipehandler.simple.php
<?php # using PHP7

require_once(__DIR__.'/output.php');

# ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----- ----
# pipe

function handle_command_event($monitor, $event)
{
    // find out event origin
    $column = explode("\t", $event);
    $origin = $column[0];

    switch($origin) {
    case 'reload':
        system('pkill lemonbar');
        break;
    case 'quit_panel':
        exit(1);
    case 'tag_changed':
    case 'tag_flags':
    case 'tag_added':
    case 'tag_removed':
        set_tag_value($monitor);
        break;
    case 'window_title_changed':
    case 'focus_changed':
        set_windowtitle($column[2]);
        break;
    }
}

function init_content($monitor, $lemon_stdin)
{   
    // initialize statusbar before loop
    set_tag_value($monitor);
    set_windowtitle('');
        
    $text = get_statusbar_text($monitor);
    fwrite($lemon_stdin, $text."\n");
    flush();
}

function walk_content($monitor, $lemon_stdin)
{       
    // start a pipe
    $descriptorspec = array(
        0 => array('pipe', 'r'),  // stdin
        1 => array('pipe', 'w'),  // stdout
        2 => array('pipe', 'w',)  // stderr
    );

    $command_in = 'herbstclient --idle';
    $proc_in  = proc_open($command_in,  $descriptorspec, $pipe_in);
    
    
    while(!feof($pipe_in[1])) {
        # read next event
        $event = fgets($pipe_in[1]);
        handle_command_event($monitor, $event);
        
        $text = get_statusbar_text($monitor);
        fwrite($lemon_stdin, $text."\n");
        flush();
    }
    
    pclose($pipe_in);
}

function run_lemon($monitor, $parameters) 
{ 
    $descriptorspec = array(
        0 => array('pipe', 'r'),  // stdin
        1 => array('pipe', 'w'),  // stdout
        2 => array('pipe', 'w',)  // stderr
    );
    
    $command_out  = "lemonbar $parameters";
    $proc_out = proc_open($command_out, $descriptorspec, $pipe_lemon);
    
    init_content($monitor, $pipe_lemon[0]);
    walk_content($monitor, $pipe_lemon[0]); // loop for each event

    pclose($pipe_lemon);
}

function detach_lemon($monitor, $parameters)
{ 
    $pid = pcntl_fork();
    
    switch($pid) {         
    case -1 : // fork errror         
        die('could not fork');
    case 0  : // we are the child
        run_lemon($monitor, $parameters); 
        break;
    default : // we are the parent             
        return $pid;
    }    
}
